### PR TITLE
add sref

### PR DESCRIPTION
**Pull request by mat-hek**

### CODE REVIEW

File: lib/bunch/sref.ex [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      iex> <<"#", hash::binary-size(8)>> = #{inspect(__MODULE__)}.new() |> inspect()
+      iex> Base.decode16(hash, case: :lower) |> elem(0)
+      :ok
+
+  """
+  @spec new(reference) :: t
+  def new(ref \\ make_ref()) do
+    '#Ref' ++ ref_list = :erlang.ref_to_list(ref)
+    <<bin_hash_part::binary-size(4), _::binary>> = :crypto.hash(:sha, ref_list)
+    hash = "#" <> Base.encode16(bin_hash_part, case: :lower, padding: false)
```
Comment (bblaszkow06) [inline lib/bunch/sref.ex:24]: encode16 doesn't have `padding` option

File: lib/bunch/sref.ex [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      iex> IEx.Helpers.ref(0, 1, 2, 3) |> #{inspect(__MODULE__)}.new() |> inspect()
+      "#6f1ef15f"
+      iex> <<"#", hash::binary-size(8)>> = #{inspect(__MODULE__)}.new() |> inspect()
+      iex> Base.decode16(hash, case: :lower) |> elem(0)
+      :ok
+
+  """
+  @spec new(reference) :: t
+  def new(ref \\ make_ref()) do
+    '#Ref' ++ ref_list = :erlang.ref_to_list(ref)
```
Comment (bblaszkow06) [inline lib/bunch/sref.ex:22]: If you only calculate hash - what's the point of splitting this charlist?

File: lib/bunch/sref.ex [hunk 3]
```diff
@@ -0,0 +1,32 @@
+defmodule Bunch.Sref do
+  @moduledoc """
+  A bunch of stuff for shortly-inspectable references.
```
Comment (bblaszkow06) [inline lib/bunch/sref.ex:3]: * one function is hardly a bunch
* this description could be used if such thing as 'shortly-inspectable references' existed. However, it's introducing that, so it should rather explain what it is

File: lib/bunch/sref.ex [hunk 4]
```diff
@@ -0,0 +1,32 @@
+defmodule Bunch.Sref do
```
Comment (bblaszkow06) [inline lib/bunch/sref.ex:1]: The name doesn't mean much. When I saw a notification with `add sref` I thought it's some format specific field in something like opus. Or secure ref. The point is: it does not describe that it is well. 
Ideas for alternatives: `Hash(ed)Ref`, `ShortRef`

File: lib/bunch/sref.ex (hunk 1 above)
Comment (bblaszkow06) [reply lib/bunch/sref.ex:24]: BTW, why encode with base 16? using base64 you would get more condensed representation

Comment (mat-hek) [reply lib/bunch/sref.ex]: I relied on the way Git displays commits' references. Also, with base64 we'd gain only two characters, while it's alphabet is harder to reason about than just hex imho

File: lib/bunch/sref.ex (hunk 2 above)
Comment (mat-hek) [reply lib/bunch/sref.ex]: it reduces the likelihood of a conflict

Comment (bblaszkow06) [reply lib/bunch/sref.ex]: That's not how hash works :P Similarity of input data doesn't imply similarity of the output.
If it did, that wouldn't be a good hash.
It is even quite common when storing hashed passwords to add a common (and secret) part called 'pepper' https://en.wikipedia.org/wiki/Pepper_(cryptography)

What's more:
```
iex(7)> make_ref |> :erlang.ref_to_list
'#Ref<0.2562459300.1782317058.164570>'
iex(8)> make_ref |> :erlang.ref_to_list
'#Ref<0.2562459300.1782317058.164577>'
iex(9)> make_ref |> :erlang.ref_to_list
'#Ref<0.2562459300.1782317060.164483>'
iex(10)> make_ref |> :erlang.ref_to_list
'#Ref<0.2562459300.1782317060.164490>'
iex(11)> make_ref |> :erlang.ref_to_list
'#Ref<0.2562459300.1782317060.164497>'
```
If what you said was true we'd had much bigger problem than `'#Ref'`

Comment (mat-hek) [reply lib/bunch/sref.ex]: > That's not how hash works :P Similarity of input data doesn't imply similarity of the output.

It's not about the similarity, it's about length. The longer the input, the bigger the space of possible inputs, the bigger chance of a conflict.

> It is even quite common when storing hashed passwords to add a common (and secret) part called 'pepper' 

'(and secret)' is the key there ;) The use of hash doesn't have anything to do with security in this case.

Comment (bblaszkow06) [reply lib/bunch/sref.ex]: > It's not about the similarity, it's about length. The longer the input, the bigger the space of possible inputs, the bigger chance of a conflict.

You're missing an important detail - if the prefix is constant, the space size of inputs doesn't change at all.
What's more, SHA works on blocks of 512-bits, so the input will be padded with 0s (up to 448 bits) and that makes stripping the prefix a complete waste of time.

File: lib/bunch/short_ref.ex [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      iex> IEx.Helpers.ref(0, 1, 2, 3) |> #{inspect(__MODULE__)}.new() |> inspect()
+      "#6f1ef15f"
+      iex> <<"#", hash::binary-size(8)>> = #{inspect(__MODULE__)}.new() |> inspect()
+      iex> Base.decode16(hash, case: :lower) |> elem(0)
+      :ok
+
+  """
+  @spec new(reference) :: t
+  def new(ref \\ make_ref()) do
+    '#Ref' ++ ref_list = :erlang.ref_to_list(ref)
```
Comment (bblaszkow06) [inline lib/bunch/short_ref.ex:42]: Resolve https://github.com/membraneframework/bunch/pull/31#discussion_r421561877

File: lib/bunch/short_ref.ex [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    <<bin_hash_part::binary-size(4), _::binary>> = :crypto.hash(:sha, ref_list)
```
Comment (bblaszkow06) [inline lib/bunch/short_ref.ex:43]: If you relied on GH then why not using seven characters?

File: lib/bunch/sref.ex (hunk 2 above)
Comment (mat-hek) [reply lib/bunch/sref.ex]: It does grow if you don't know the prefix is constant, as SHA algorithm doesn't. However, if it works on blocks as you said, I agree

File: lib/bunch/short_ref.ex (hunk 6 above)
Comment (mat-hek) [reply lib/bunch/short_ref.ex:43]: Because that would involve matching on 3.5 bits in that line and that wouldn't be handled by Base.encode16

**Comment by mat-hek** [comment]
> Still, you haven't provided a justification for adding this. Why would anyone use this outside Membrane?

Provided one in the moduledoc. It doesn't mention Membrane anywhere.

File: lib/bunch/short_ref.ex (hunk 6 above)
Comment (bblaszkow06) [reply lib/bunch/short_ref.ex:43]: > 3.5 bits
Bytes :P You could encode first and then take 7 characters, but never mind

Comment (mat-hek) [reply lib/bunch/short_ref.ex:43]: yes, bytes :P

**Pull request merged**